twister: Make --list-tests respect tag filter

decsny opened the pull request:
Make the --list-tests option output respect the --tag options, so that only the tests for the specified tags are listed.

Commit:
twister: Make --list-tests respect tag filter

Make the --list-tests and --tests-tree options outputs
respect the --tag and --exclude-tag options,
so that only the tests for the specified tags are listed.

Also update the TestPlan report testcases for this change.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
@nashif Not super familiar with twister and it's use, so don't actually know if this is intended behavior or not

hakehuang commented:
@decsny  as you change the  `--list-tags` behaviour, you need update the testcase to reflect this change as well

nashif commented:
> @nashif Not super familiar with twister and it's use, so don't actually know if this is intended behavior or not

`--list-tags ` gives you the list of all tags, `--tag` filters by a specific tag, so not sure what are you trying to do here, can you elaborate on the usecase?

decsny commented:
> > @nashif Not super familiar with twister and it's use, so don't actually know if this is intended behavior or not
> 
> `--list-tags ` gives you the list of all tags, `--tag` filters by a specific tag, so not sure what are you trying to do here, can you elaborate on the usecase?

Sorry I just realized I made a typo in the PR title and description, it's actually changing --list-tests

decsny commented:
The commit description is also wrong I will fix that next time I push

henrikbrixandersen commented:
> For me works good, I tried with command like: `./zephyr/scripts/twister -c -T zephyr/tests/boot --tag recovery --tag pytest --list-tests` and as a result I see only tests with given tags. However:
> 
> * I thought that `--list-tests` and `--test-tree` are easy commands just to show all tests (without using any filtering)
> * if you use filters, why not apply another, like `--exclude-tag`, `--enable-slow` ...
> * and such a filter should also be applied to `--test-tree` command
> 
> To see the final list I normally use `--save-tests` command, then you receive full list of configurations (with platform) in yaml file (copy of testplan.yaml).

Being able to list the tests matching a given tag is useful, I think. I often want to run e.g. a specific CAN test, but instead of digging through either the process above or looking through YAML files for finding the identifier of the given test, simply invoking `twister --list-tests -t can` would be much easier.

decsny commented:
> * I thought that `--list-tests` and `--test-tree` are easy commands just to show all tests (without using any filtering)

not sure if I understand this objection because this is still preserved by this PR, if you do --list-tests with no --tag then it will list all the tests.

gchwier commented:
> > * I thought that `--list-tests` and `--test-tree` are easy commands just to show all tests (without using any filtering)
> 
> not sure if I understand this objection because this is still preserved by this PR, if you do --list-tests with no --tag then it will list all the tests.

I agree. I shouldn't add this objection, rather I thought about other filters, why not apply them ... but it is not so easy and it could raise another problems. Filtering only tags is also very useful.
> 
> > and such a filter should also be applied to --test-tree command
> 
> I'll look into this

You can just replace `get_all_tests` with:
```
    def get_tests_list(self):
        testcases = []
        if tag_filter := self.options.tag:
            for _, ts in self.testsuites.items():
                if ts.tags.intersection(tag_filter):
                    for case in ts.testcases:
                        testcases.append(case.name)
        else:
            for _, ts in self.testsuites.items():
                for case in ts.testcases:
                    testcases.append(case.name)

        if exclude_tag := self.options.exclude_tag:
            for _, ts in self.testsuites.items():
                if ts.tags.intersection(exclude_tag):
                    for case in ts.testcases:
                        if case.name in testcases:
                            testcases.remove(case.name)
        return testcases
```
I added also filtering for `--exclude-tag`.  And use that new method instead of `get_all_tests`:
```
    def report_test_tree(self):
        all_tests = self.get_tests_list()
       ...

    def report_test_list(self):
        all_tests = self.get_tests_list()
        ...
```

Tested with commands:
```
./scripts/twister -T tests/boot --tag sysbuild --tag pytest --exclude-tag retention --list-tests
 - boot.with_mcumgr.test_downgrade_prevention
 - boot.with_mcumgr.test_upgrade
 - bootloader.mcuboot.recovery.retention
 - bootloader.mcuboot.recovery.retention.mem
4 total.

./scripts/twister -T tests/boot --tag sysbuild --tag pytest --exclude-tag retention --test-tree
Testsuite
├── Samples
└── Tests
    ├── boot
    │   └── with_mcumgr
    │       ├── boot.with_mcumgr.test_downgrade_prevention
    │       └── boot.with_mcumgr.test_upgrade
    └── bootloader
        └── mcuboot
            ├── bootloader.mcuboot.recovery.retention
            └── bootloader.mcuboot.recovery.retention.mem
```

decsny commented:
> > > * I thought that `--list-tests` and `--test-tree` are easy commands just to show all tests (without using any filtering)
> > 
> > 
> > not sure if I understand this objection because this is still preserved by this PR, if you do --list-tests with no --tag then it will list all the tests.
> 
> I agree. I shouldn't add this objection, rather I thought about other filters, why not apply them ... but it is not so easy and it could raise another problems. Filtering only tags is also very useful.
> 
> > > and such a filter should also be applied to --test-tree command
> > 
> > 
> > I'll look into this
> 
> You can just replace `get_all_tests` with:
> 
> ```
>     def get_tests_list(self):
>         testcases = []
>         if tag_filter := self.options.tag:
>             for _, ts in self.testsuites.items():
>                 if ts.tags.intersection(tag_filter):
>                     for case in ts.testcases:
>                         testcases.append(case.name)
>         else:
>             for _, ts in self.testsuites.items():
>                 for case in ts.testcases:
>                     testcases.append(case.name)
> 
>         if exclude_tag := self.options.exclude_tag:
>             for _, ts in self.testsuites.items():
>                 if ts.tags.intersection(exclude_tag):
>                     for case in ts.testcases:
>                         if case.name in testcases:
>                             testcases.remove(case.name)
>         return testcases
> ```
> 
> I added also filtering for `--exclude-tag`. And use that new method instead of `get_all_tests`:
> 
> ```
>     def report_test_tree(self):
>         all_tests = self.get_tests_list()
>        ...
> 
>     def report_test_list(self):
>         all_tests = self.get_tests_list()
>         ...
> ```
> 
> Tested with commands:
> 
> ```
> ./scripts/twister -T tests/boot --tag sysbuild --tag pytest --exclude-tag retention --list-tests
>  - boot.with_mcumgr.test_downgrade_prevention
>  - boot.with_mcumgr.test_upgrade
>  - bootloader.mcuboot.recovery.retention
>  - bootloader.mcuboot.recovery.retention.mem
> 4 total.
> 
> ./scripts/twister -T tests/boot --tag sysbuild --tag pytest --exclude-tag retention --test-tree
> Testsuite
> ├── Samples
> └── Tests
>     ├── boot
>     │   └── with_mcumgr
>     │       ├── boot.with_mcumgr.test_downgrade_prevention
>     │       └── boot.with_mcumgr.test_upgrade
>     └── bootloader
>         └── mcuboot
>             ├── bootloader.mcuboot.recovery.retention
>             └── bootloader.mcuboot.recovery.retention.mem
> ```

okay, I took this suggestion, thanks

also @hakehuang  I updated the twister tests, thanks
